hw_diag/utilities/network_watchdog.py

```diff
@@ -0,0 +1,146 @@
+import logging
+import os
+import tempfile
+from datetime import datetime, timedelta
+from logging.handlers import RotatingFileHandler
+import socket
+
+from hm_pyhelper.logger import get_logger
+from hw_diag.utilities.balena_supervisor import BalenaSupervisor
+from hw_diag.utilities.dbus_proxy.dbus_ids import DBusIds
+from hw_diag.utilities.dbus_proxy.network_manager import NetworkManager
+from hw_diag.utilities.dbus_proxy.systemd import Systemd
+from hw_diag.utilities.keystore import KeyStore
+
+
+class NetworkWatchdog:
+    VOLUME_PATH = '/var/watchdog/'
+    WATCHDOG_LOG_FILE_NAME = 'watchdog.log'
+    LAST_RESTART_FILE_NAME = 'last_restart.json'
+
+    MAX_LOG_SIZE = 10 * 1024 * 1024  # 10 Mb
+    LAST_RESTART_KEY = 'last_restart'
+    LAST_RESTART_DATE_FORMAT = '%d/%m/%Y %H:%M:%S'
+
+    # Full system reboot limited to once a day
+    REBOOT_LIMIT_HOURS = 24
+    # Failed connectivity count for the network manager to restart
+    NM_RESTART_THRESHOLD = int(os.environ.get("NM_RESTART_THRESHOLD", 1))
+    # Failed connectivity count for the hotspot to reboot
+    FULL_REBOOT_THRESHOLD = int(os.environ.get("FULL_REBOOT_THRESHOLD", 3))
+    # Failed connectivity count for the hotspot to reboot
+    FULL_FORCE_REBOOT_THRESHOLD = int(os.environ.get("FULL_FORCE_REBOOT_THRESHOLD", 6))
+
+    # Public DNS server for checking internet connectivity
+    PUBLIC_DNS_SERVER = "8.8.8.8"       # NOSONAR
+    PUBLIC_DNS_PORT = 53
+
+    # Static variable for saving the lost connectivity count
+    lost_count = 0
+
+    def __init__(self):
+        if os.access(self.VOLUME_PATH, os.W_OK):
+            self.log_file_path = os.path.join(self.VOLUME_PATH, self.WATCHDOG_LOG_FILE_NAME)
+            self.state_file_path = os.path.join(self.VOLUME_PATH, self.LAST_RESTART_FILE_NAME)
+        else:
+            self.temp_dir = tempfile.TemporaryDirectory()
+            self.log_file_path = os.path.join(self.temp_dir.name, self.WATCHDOG_LOG_FILE_NAME)
+            self.state_file_path = os.path.join(self.temp_dir.name, self.LAST_RESTART_FILE_NAME)
+
+        # Set up logger
+        self.logger = get_logger(__name__)
+        # Add rotating log handler to the logger
+        handler = RotatingFileHandler(self.log_file_path, maxBytes=self.MAX_LOG_SIZE, backupCount=3)
+        handler.setLevel(logging.INFO)
+        handler.setFormatter(
+            logging.Formatter("%(asctime)s - [%(levelname)s]:(%(lineno)d) - %(message)s"))
```

Comment:
Actually this log format stems from `pyhelper` and I simplified it since all watchdog logs come from the single python file. 
However for the consistency, I updated to use the same logging format with `pyhelper`.

Reply:
Simple logging format for watchdog("%(asctime)s - [%(levelname)s]:(%(lineno)d) - %(message)s"):
```
2022-06-06 11:04:45,076 - [INFO]:(64) - Watchdog is started!
2022-06-06 11:04:45,083 - [INFO]:(109) - Running the watchdog...
2022-06-06 11:04:45,090 - [INFO]:(81) - Checking the network connectivity.
2022-06-06 11:04:45,101 - [INFO]:(84) - Network manager state: Connected-Global
2022-06-06 11:04:45,264 - [INFO]:(85) - Internet connectivity: True
2022-06-06 11:04:45,273 - [INFO]:(113) - Network is working.
```

Use `pyhelper` logging format for watchdog:
```
2022-06-06 16:19:44,225 - [INFO] - hw_diag.utilities.network_watchdog - (network_watchdog.py).__init__ -- /opt/python-dependencies/hw_diag/utilities/network_watchdog.py:(83) - Watchdog is started!
2022-06-06 16:19:44,230 - [INFO] - hw_diag.utilities.network_watchdog - (network_watchdog.py).ensure_network_connection -- /opt/python-dependencies/hw_diag/utilities/network_watchdog.py:(128) - Running the watchdog...
2022-06-06 16:19:44,234 - [INFO] - hw_diag.utilities.network_watchdog - (network_watchdog.py).is_connected -- /opt/python-dependencies/hw_diag/utilities/network_watchdog.py:(100) - Checking the network connectivity.
2022-06-06 16:19:44,246 - [INFO] - hw_diag.utilities.network_watchdog - (network_watchdog.py).is_connected -- /opt/python-dependencies/hw_diag/utilities/network_watchdog.py:(103) - Network manager state: Connected-Global
2022-06-06 16:19:44,417 - [INFO] - hw_diag.utilities.network_watchdog - (network_watchdog.py).is_connected -- /opt/python-dependencies/hw_diag/utilities/network_watchdog.py:(104) - Internet connectivity: True
2022-06-06 16:19:44,426 - [INFO] - hw_diag.utilities.network_watchdog - (network_watchdog.py).ensure_network_connection -- /opt/python-dependencies/hw_diag/utilities/network_watchdog.py:(133) - Network is working.
```

@marvinmarnold As the network watchdog log format, what about using the simple logging format so that we can focus on the network status without distracting to the location of the log statement?

Reply:
No strong opinion. The changes look good to me but I'll let @pritamghanghas officially approve the PR when he's ready.